### PR TITLE
security v1

### PR DESCRIPTION
**Pull request by alexusievich**

### CODE REVIEW

File: src/main/java/com/mycompany/shop/basket/BasketController.java [hunk 1]
```diff
@@ -33,21 +39,75 @@
...
     @GetMapping
+    public ResponseEntity<Basket> getUserBasket(@RequestParam("userId") Long userId) {
+        Basket basket;
+        if (userId != null) {
+            if (basketRepository.findBasketByUserId(userId).isPresent()) {
+                basket = basketRepository.findBasketByUserId(userId).get();
+                sessionBean.setBasket(basket);
+            } else {
+                sessionBean.setBasket(null);
+            }
```
Comment (nikmonkov) [inline src/main/java/com/mycompany/shop/basket/BasketController.java:70]: that logic should be done on login step. also here I can replace userId by any userId, so I can get basket of any user in application

File: src/main/java/com/mycompany/shop/basket/BasketController.java [hunk 2]
```diff
@@ -33,21 +39,75 @@
...
+    }
+
+    @GetMapping("/unregistered")
     public ResponseEntity<Basket> getBasket() {
         return ResponseEntity.ok(sessionBean.getBasket());
     }
 
     @PostMapping
+    public ResponseEntity<Basket> addToUserBasket(@RequestBody CreateBasketRequest createBasketRequest) {
+        Long userId = createBasketRequest.getUserId();
```
Comment (nikmonkov) [inline src/main/java/com/mycompany/shop/basket/BasketController.java:82]: userId should be retrieved from the session

File: src/main/java/com/mycompany/shop/basket/BasketController.java [hunk 3]
```diff
@@ -62,14 +122,22 @@ public void setId(Long id) {
...
         }
     }
 
+    @DeleteMapping("/logout")
+    public void clearBasketFromSession() {
+        Basket basket = sessionBean.getBasket();
+        if (basket != null) {
+            sessionBean.setBasket(null);
+        }
+    }
```
Comment (nikmonkov) [inline src/main/java/com/mycompany/shop/basket/BasketController.java:138]: why here? it should be part of general logout method

File: src/main/java/com/mycompany/shop/security/SecurityController.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        UserDetailsImpl userDetails = (UserDetailsImpl) authentication.getPrincipal();
+
+        User authenticatedUser = new User(userDetails.getId(), userDetails.getUsername());
+
+        return ResponseEntity.ok(authenticatedUser);
+    }
+
+    @PostMapping("/logout")
+    public void logout() {
+        SecurityContextHolder.clearContext();
```
Comment (nikmonkov) [inline src/main/java/com/mycompany/shop/security/SecurityController.java:51]: here you should clear the basket from the session

File: src/main/java/com/mycompany/shop/security/SecurityController.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        Authentication authentication = authenticationManager.authenticate(
+                new UsernamePasswordAuthenticationToken(user.getUsername(), user.getPassword()));
+
+        SecurityContextHolder.getContext().setAuthentication(authentication);
+
+        UserDetailsImpl userDetails = (UserDetailsImpl) authentication.getPrincipal();
+
+        User authenticatedUser = new User(userDetails.getId(), userDetails.getUsername());
+
+        return ResponseEntity.ok(authenticatedUser);
```
Comment (nikmonkov) [inline src/main/java/com/mycompany/shop/security/SecurityController.java:46]: here you need to get basket for this user and keep it in the session bean

File: src/main/java/com/mycompany/shop/security/SecurityController.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+            userRepository.save(newUser);
+            return ResponseEntity.ok(newUser.getUsername());
+        }
+    }
+
+
+
+
+
+
```
Comment (nikmonkov) [inline src/main/java/com/mycompany/shop/security/SecurityController.java:74]: remove empty lines. chech the rest of the files also

File: src/main/java/com/mycompany/shop/security/userdetails/UserDetailsServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+import com.mycompany.shop.user.User;
+import com.mycompany.shop.user.UserRepository;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.security.core.userdetails.UserDetails;
+import org.springframework.security.core.userdetails.UserDetailsService;
+import org.springframework.security.core.userdetails.UsernameNotFoundException;
+import org.springframework.stereotype.Service;
+
+
+@Service("userDetailsServiceImpl")
```
Comment (nikmonkov) [inline src/main/java/com/mycompany/shop/security/userdetails/UserDetailsServiceImpl.java:13]: do you actually need to specify "userDetailsServiceImpl" in annotation?

File: src/main/java/com/mycompany/shop/user/UserController.java [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+@RestController
+@RequestMapping("/api/userinfo")
+public class UserController {
+
+    @Autowired
+    UserRepository userRepository;
+
+    @GetMapping("/{id}")
+    public ResponseEntity<User> retrieveProduct(@PathVariable long id) {
```
Comment (nikmonkov) [inline src/main/java/com/mycompany/shop/user/UserController.java:18]: why method called "retrieveProduct"?

File: src/main/java/com/mycompany/shop/user/UserController.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+        Optional<User> optionalUser = userRepository.findById(id);
+
+        if (optionalUser.isEmpty()) {
+            throw new ProductNotFoundException("The user with id: " + id + " is not found");
```
Comment (nikmonkov) [inline src/main/java/com/mycompany/shop/user/UserController.java:23]: why "ProductNotFoundException"?

File: src/main/java/com/mycompany/shop/basket/BasketController.java [hunk 10]
```diff
@@ -15,9 +18,12 @@
 
 
 @RestController
-@RequestMapping("/api/basket")
+@RequestMapping("/api/public/basket")
```
Comment (nikmonkov) [inline src/main/java/com/mycompany/shop/basket/BasketController.java:21]: I don't like the idea of adding this additional word "public" in all APIs. better to add additional work only for private API

File: src/main/java/com/mycompany/shop/basket/BasketController.java [hunk 11]
```diff
@@ -33,21 +39,75 @@
...
+        return ResponseEntity.ok(sessionBean.getBasket());
+    }
+
+    @GetMapping("/unregistered")
     public ResponseEntity<Basket> getBasket() {
         return ResponseEntity.ok(sessionBean.getBasket());
     }
 
     @PostMapping
+    public ResponseEntity<Basket> addToUserBasket(@RequestBody CreateBasketRequest createBasketRequest) {
```
Comment (nikmonkov) [inline src/main/java/com/mycompany/shop/basket/BasketController.java:81]: you don't need additional rest method for adding into basket for authorized user. use one method and build the logic inside the method based on information from session if user is logged in or not

File: src/main/webapp/src/App.js [hunk 12]
```diff
@@ -23,53 +27,144 @@ class App extends React.Component {
...
+                icon: <LogoutOutlined style={{color: '#108ee9', fontSize: 30}}/>,
+            })
+            this.setState({user: null});
+            this.setState({currentUser: null});
+            Cookie.deleteCookie("user");
+        });
+        axios.delete("/api/public/basket/logout").then(response => {
+                this.setState({basket: null});
             }
         );
```
Comment (nikmonkov) [inline src/main/webapp/src/App.js:125]: use one single call for logout.

File: src/main/webapp/src/App.js [hunk 13]
```diff
@@ -23,53 +27,144 @@ class App extends React.Component {
...
     }
 
+    submitForm = (username, password) => {
+        axios.post("/api/public/auth/login", {username: username, password: password}).then(response => {
+            this.setState({currentUser: response.data});
+            if (this.state.currentUser) {
+                Cookie.setCookie("user", JSON.stringify(this.state.currentUser));
+            }
```
Comment (nikmonkov) [inline src/main/webapp/src/App.js:133]: do you actually need this info in cookies? why do you need userId on client?

**Review by nikmonkov — CHANGES_REQUESTED** [review]

File: src/main/java/com/mycompany/shop/basket/BasketController.java (hunk 1 above)
Comment (alexusievich) [reply src/main/java/com/mycompany/shop/basket/BasketController.java:70]: +

File: src/main/java/com/mycompany/shop/basket/BasketController.java (hunk 3 above)
Comment (alexusievich) [reply src/main/java/com/mycompany/shop/basket/BasketController.java:138]: +

File: src/main/java/com/mycompany/shop/security/SecurityController.java (hunk 5 above)
Comment (alexusievich) [reply src/main/java/com/mycompany/shop/security/SecurityController.java:46]: +

File: src/main/webapp/src/App.js (hunk 12 above)
Comment (alexusievich) [reply src/main/webapp/src/App.js:125]: +

File: src/main/java/com/mycompany/shop/basket/BasketController.java (hunk 11 above)
Comment (alexusievich) [reply src/main/java/com/mycompany/shop/basket/BasketController.java:81]: +

File: src/main/webapp/src/App.js (hunk 13 above)
Comment (alexusievich) [reply src/main/webapp/src/App.js:133]: +

Comment (alexusievich) [reply src/main/webapp/src/App.js:133]: +

File: src/main/java/com/mycompany/shop/basket/BasketController.java (hunk 10 above)
Comment (alexusievich) [reply src/main/java/com/mycompany/shop/basket/BasketController.java:21]: +

File: src/main/webapp/src/components/Basket.js [hunk 14]
```diff
@@ -49,7 +49,9 @@ class Basket extends React.Component {
                 <div>
                     <div className="basketTitle">
                         <div className="basketName">Shopping cart - {this.props.basket.items.length} items</div>
-                        <div className="basketTotal">Total: {this.props.basket.totalPrice / 1000}.000 RUB</div>
+                        <div className="basketTotal">
+                            Total: {Math.trunc(this.props.basket.totalPrice / 1000)} {this.props.basket.totalPrice % 1000 === 0 ? "000" : this.props.basket.totalPrice % 1000} RUB
```
Comment (nikmonkov) [inline src/main/webapp/src/components/Basket.js:53]: rewrite

File: src/main/webapp/src/components/ProductDetails.js [hunk 15]
```diff
@@ -86,13 +86,13 @@ class ProductDetails extends React.Component {
...
                         <div className="rating">
-                            {this.state.phoneDetails.rating} <StarFilled style={{color: "#1890ff"}}/>
+                            <StarFilled style={{color: "#1890ff"}}/> {this.state.phoneDetails.rating}
                         </div>
                         <div className="description">
                             {this.state.phoneDetails.description}
                         </div>
                         <div className="pricee">
-                            {this.state.phoneDetails.price} RUB
+                            {this.state.phoneDetails.price/1000} 000 RUB
```
Comment (nikmonkov) [inline src/main/webapp/src/components/ProductDetails.js:95]: rewrite

File: src/main/webapp/src/components/Products.js [hunk 16]
```diff
@@ -24,11 +24,10 @@ class Products extends React.Component {
             current: 1,
             searchTerm: '',
             sortName: 'Sorting',
-            minPrice: 0,
-            maxPrice: 0,
+            minPrice: undefined,
+            maxPrice: undefined,
```
Comment (nikmonkov) [inline src/main/webapp/src/components/Products.js:28]: set null and check if it works

File: src/main/webapp/src/components/Products.js [hunk 17]
```diff
@@ -225,7 +232,7 @@ class Products extends React.Component {
                         description={product.config}
                     />
                     <div className="priceCart">
-                        <div className="price">{product.price} RUB</div>
+                        <div className="price">{product.price >= 1000 ? product.price/1000 + " 000" : product.price} RUB</div>
```
Comment (nikmonkov) [inline src/main/webapp/src/components/Products.js:235]: rewrite

File: src/main/webapp/src/App.js [hunk 18]
```diff
@@ -23,34 +26,46 @@ class App extends React.Component {
...
             this.setState({basket});
         })
+        this.state.currentUser && axios.get("/api/userinfo/" + this.state.currentUser.id).then(response => {
+                const user = response.data;
+                this.setState({user});
+            }
+        )
     };
 
+    getBasket = (response, name) => {
```
[the remark this answers is not in the document]
Comment (alexusievich) [reply src/main/webapp/src/App.js:46]: +

File: src/main/webapp/src/components/Basket.js (hunk 14 above)
Comment (alexusievich) [reply src/main/webapp/src/components/Basket.js:53]: +

File: src/main/webapp/src/components/ProductDetails.js (hunk 15 above)
Comment (alexusievich) [reply src/main/webapp/src/components/ProductDetails.js:95]: +

File: src/main/webapp/src/components/Products.js (hunk 17 above)
Comment (alexusievich) [reply src/main/webapp/src/components/Products.js:235]: +

**Review by nikmonkov — APPROVED** [review]

**Pull request merged**